File: android/sample-ui/src/main/kotlin/com/usesmileid/sampleapps/ui/components/UseSmileIDSampleJobRow.kt
package com.usesmileid.sampleapps.ui.components

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ExperimentalLayoutApi
import androidx.compose.foundation.layout.FlowRow
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import com.smileid.designsystem.SmileDimens
import com.usesmileid.sampleapps.ui.UseSmileIDSampleTestIds
import com.usesmileid.sampleapps.ui.model.UseSmileIDSampleProduct
import com.usesmileid.sampleapps.ui.theme.UseSmileIDSampleTheme

/**
 * One verification: a product tile, its name, a secondary line of job id and time, and the status badge.
 *
 * Select mode's checkbox is not a slot — the design puts it beside the card, and inside it cost the title width.
 *
 * @param statusTestId defaulted rather than attached by the caller; a list screen overrides it with the row's suffix.
 */
@OptIn(ExperimentalLayoutApi::class)
@Composable
fun UseSmileIDSampleJobRow(
    product: UseSmileIDSampleProduct,
    jobId: String,
    time: String,
    status: UseSmileIDSampleStatus,
    modifier: Modifier = Modifier,
    onClick: (() -> Unit)? = null,
    testId: String? = null,
    statusTestId: String? = UseSmileIDSampleTestIds.JOB_ROW_STATUS,
) {
    val colors = UseSmileIDSampleTheme.colors
    Surface(
        modifier = modifier
            .fillMaxWidth()
            .then(if (onClick != null) Modifier.clickable(role = Role.Button, onClick = onClick) else Modifier)
            .tagged(testId),
        shape = RoundedCornerShape(SmileDimens.radiusSurface),
        color = colors.card.background,
        border = BorderStroke(SmileDimens.borderWidthHairline, colors.card.border),
    ) {
        // A Row at the design's scale keeps the badge inline; a FlowRow above it lets the badge drop.
        // One layout cannot do both: weight() inside a FlowRow claims the whole line.
        val stacks = LocalDensity.current.fontScale > 1f
        val padding = Modifier
            .defaultMinSize(minHeight = SmileDimens.space64)
            .padding(SmileDimens.spacingSm)
        if (stacks) {
            FlowRow(
                modifier = padding,
                horizontalArrangement = Arrangement.spacedBy(SmileDimens.spacingSm),
                verticalArrangement = Arrangement.spacedBy(SmileDimens.spacingXs),
                itemVerticalAlignment = Alignment.CenterVertically,
            ) {
                JobRowTile(product = product)
                JobRowText(product = product, jobId = jobId, time = time, stacks = true)
                UseSmileIDSampleStatusBadge(status = status, testId = statusTestId)
            }
        } else {
            Row(
                modifier = padding,
                horizontalArrangement = Arrangement.spacedBy(SmileDimens.spacingSm),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                JobRowTile(product = product)
                JobRowText(
                    product = product,
                    jobId = jobId,
                    time = time,
                    stacks = false,
                    modifier = Modifier.weight(1f),
                )
                UseSmileIDSampleStatusBadge(status = status, testId = statusTestId)
            }
        }
    }
}

@Composable
private fun JobRowTile(product: UseSmileIDSampleProduct) {
    val hue = product.hue
    Surface(
        modifier = Modifier.size(TILE_SIZE),
        shape = RoundedCornerShape(TILE_RADIUS),
        color = hue.tile,
    ) {
        Box(contentAlignment = Alignment.Center) {
            val icon = product.iconRes
            if (icon != null) {
                UseSmileIDSampleIcon(id = icon, tint = hue.icon, size = TILE_ICON_SIZE)
            } else {
                ProductMarkGlyph(tint = hue.icon)
            }
        }
    }
}

@Composable
private fun JobRowText(
    product: UseSmileIDSampleProduct,
    jobId: String,
    time: String,
    stacks: Boolean,
    modifier: Modifier = Modifier,
) {
    val colors = UseSmileIDSampleTheme.colors
    Column(modifier = modifier, verticalArrangement = Arrangement.spacedBy(SmileDimens.spacingXxs)) {
        Text(
            text = product.label,
            style = UseSmileIDSampleTheme.type.textStyleBodyStrong,
            color = colors.card.title,
            // One line at the design's scale; enlarged type wraps, because eliding it would clip.
            maxLines = if (stacks) Int.MAX_VALUE else 1,
            overflow = TextOverflow.Ellipsis,
        )
        Text(
            text = "$jobId · $time",
            style = UseSmileIDSampleTheme.type.textStyleBodySm,
            color = colors.textMuted,
            // Ellipsised like the title, so every row is the same height at the design's scale.
            maxLines = if (stacks) Int.MAX_VALUE else 1,
            overflow = TextOverflow.Ellipsis,
        )
    }
}

private val TILE_SIZE = 36.dp
private val TILE_RADIUS = 10.dp
private val TILE_ICON_SIZE = 18.dp

File: flutter/sample_ui/lib/src/components/use_smileid_sample_job_row.dart
import 'package:flutter/material.dart';

import '../model/use_smileid_sample_product.dart';
import '../model/use_smileid_sample_status.dart';
import '../theme/use_smileid_sample_colors.dart';
import '../theme/use_smileid_sample_theme.dart';
import '../theme/use_smileid_sample_typography.dart';
import '../tokens/smile_product_hues.dart';
import '../tokens/smile_tokens.dart';
import '../use_smileid_sample_test_ids.dart';
import 'use_smileid_sample_icon.dart';
import 'use_smileid_sample_status_badge.dart';

/// One verification: a product tile, its name, a secondary line of job id and time, and the badge.
///
/// Select mode's checkbox is not a slot — the design puts it beside the card, and inside it cost
/// the title its width.
class UseSmileIDSampleJobRow extends StatelessWidget {
  /// [statusTestId] is defaulted rather than attached by the caller; a list overrides it per row.
  const UseSmileIDSampleJobRow({
    required this.product,
    required this.jobId,
    required this.time,
    required this.status,
    this.onTap,
    this.testId,
    this.statusTestId = UseSmileIDSampleTestIds.jobRowStatus,
    super.key,
  });

  /// Which product the job ran.
  final UseSmileIDSampleProduct product;

  /// The job id, already elided by the caller.
  final String jobId;

  /// The submission time, already formatted.
  final String time;

  /// The job's status.
  final UseSmileIDSampleStatus status;

  /// What a tap does.
  final VoidCallback? onTap;

  /// The `sample_*` id the screen supplies.
  final String? testId;

  /// The `sample_*` id for the badge, asserted on separately.
  final String? statusTestId;

  @override
  Widget build(BuildContext context) {
    final UseSmileIDSampleColors colors = UseSmileIDSampleTheme.colorsOf(
      context,
    );
    // A Row at the design's scale keeps the badge inline; a Wrap above it lets the badge drop.
    // One layout cannot do both: a flexible child inside a Wrap claims the whole line.
    final bool stacks = MediaQuery.textScalerOf(context).scale(1) > 1;
    final Widget tile = _JobRowTile(product: product);
    final Widget badge = UseSmileIDSampleStatusBadge(
      status: status,
      testId: statusTestId,
    );

    return Semantics(
      identifier: testId,
      button: onTap != null,
      child: Material(
        color: colors.card.background,
        shape: RoundedRectangleBorder(
          borderRadius: UseSmileIDSampleShapes.card,
          side: BorderSide(
            color: colors.cardStroke,
            width: smileCardStrokeWidth,
          ),
        ),
        child: InkWell(
          onTap: onTap,
          borderRadius: UseSmileIDSampleShapes.card,
          child: ConstrainedBox(
            constraints: const BoxConstraints(minHeight: SmileDimens.space64),
            child: Padding(
              padding: const EdgeInsets.all(SmileDimens.spacingSm),
              child: stacks
                  ? Wrap(
                      spacing: SmileDimens.spacingSm,
                      runSpacing: SmileDimens.spacingXs,
                      crossAxisAlignment: WrapCrossAlignment.center,
                      children: <Widget>[
                        tile,
                        _JobRowText(
                          product: product,
                          jobId: jobId,
                          time: time,
                          stacks: true,
                          colors: colors,
                        ),
                        badge,
                      ],
                    )
                  : Row(
                      children: <Widget>[
                        tile,
                        const SizedBox(width: SmileDimens.spacingSm),
                        Expanded(
                          child: _JobRowText(
                            product: product,
                            jobId: jobId,
                            time: time,
                            stacks: false,
                            colors: colors,
                          ),
                        ),
                        const SizedBox(width: SmileDimens.spacingSm),
                        badge,
                      ],
                    ),
            ),
          ),
        ),
      ),
    );
  }
}

class _JobRowTile extends StatelessWidget {
  const _JobRowTile({required this.product});

  final UseSmileIDSampleProduct product;

  @override
  Widget build(BuildContext context) {
    final SmileProductHue hue = productHue(product);
    return Container(
      width: _tileSize,
      height: _tileSize,
      alignment: Alignment.center,
      decoration: BoxDecoration(
        color: hue.tile,
        borderRadius: UseSmileIDSampleShapes.rowTile,
      ),
      child: UseSmileIDSampleIcon(
        asset: productIcon(product),
        tint: hue.icon,
        size: _tileIconSize,
      ),
    );
  }
}

class _JobRowText extends StatelessWidget {
  const _JobRowText({
    required this.product,
    required this.jobId,
    required this.time,
    required this.stacks,
    required this.colors,
  });

  final UseSmileIDSampleProduct product;
  final String jobId;
  final String time;
  final bool stacks;
  final UseSmileIDSampleColors colors;

  @override
  Widget build(BuildContext context) => Column(
    crossAxisAlignment: CrossAxisAlignment.start,
    mainAxisSize: MainAxisSize.min,
    children: <Widget>[
      Text(
        product.label,
        // One line at the design's scale; enlarged type wraps, because eliding it would clip.
        maxLines: stacks ? null : 1,
        overflow: stacks ? null : TextOverflow.ellipsis,
        style: UseSmileIDSampleType.textStyleBodyStrong.copyWith(
          color: colors.card.title,
        ),
      ),
      const SizedBox(height: SmileDimens.spacingXxs),
      Text(
        '$jobId · $time',
        // Elided like the title, so every row is the same height at the design's scale.
        maxLines: stacks ? null : 1,
        overflow: stacks ? null : TextOverflow.ellipsis,
        style: UseSmileIDSampleType.textStyleBodySm.copyWith(
          color: colors.textMuted,
        ),
      ),
    ],
  );
}

/// The design's tile and glyph, both read off the board rather than the scale.
const double _tileSize = 36;
const double _tileIconSize = 18;
